src/haproxy.py

```diff
@@ -46,3 +73,25 @@ def is_active(self) -> bool:
             True if the haproxy is running.
         """
         return systemd.service_running(APT_PACKAGE_NAME)
+
+    def _render_file(self, path: Path, content: str, mode: int) -> None:
+        """Write a content rendered from a template to a file.
+
+        Args:
+            path: Path object to the file.
+            content: the data to be written to the file.
+            mode: access permission mask applied to the
+              file using chmod (e.g. 0o640).
+        """
+        path.write_text(content, encoding="utf-8")
+        os.chmod(path, mode)
+        u = pwd.getpwnam(HAPROXY_USER)
+        # Set the correct ownership for the file.
+        os.chown(path, uid=u.pw_uid, gid=u.pw_gid)
+
+    def _render_haproxy_config(self) -> None:
+        """Render the haproxy configuration file."""
+        with open("templates/haproxy.cfg.j2", "r", encoding="utf-8") as file:
```

Comment:
I think we can use Jinja's loaders :) https://jinja.palletsprojects.com/en/3.0.x/api/?highlight=filesystemloader#jinja2.FileSystemLoader

Reply:
Also it seems that this file does not contain any template replacement, could this be just a cfg file?

Reply:
It does look a bit weird to have a jinja2 without any replacements but I'm adding them in the following PRs :smile: so I think the file can be left as-is for now.